Use the new dag api and use ipld links for both heads and next pointers

satazor opened the pull request:
## Description (Required)

Use the new dag api and use ipld links for both heads and next pointers

The object api is deprecated and might be removed in future releases.
Additionally, we may now use https://explore.ipld.io/ to explore the log & entries!

With this change, we are now dealing with CIDs instead of multihashes. Therefore, this PR introduced `fromCID` and `toCID`. The `fromMultihash` and `toMultihash` became deprecated.
Moreover, all occurrences of `hash` were replaced to `cid` in both source code and tests.

## Other changes (e.g. bug fixes, UI tweaks, refactors)

Updated IPFS to latest version as the current installed one has breaking changes to the dag api.
This way, we are already using the new dag api.

## TODO

- [x] Add unit test
- [ ] Add documentation

Closes #200, #106

aphelionz commented:
Just saw this and haven't checked myself but is the performance issue mitigated? If so, what was it?

shamb0t commented:
@aphelionz the perceived performance hit was due to monkey-patching [here](https://github.com/orbitdb/ipfs-log/blob/1e156ddaa3451e7a0bb1ba532e031211ef28c2bf/test/replicate.spec.js#L49) (using memstore to avoid hitting disk) and the new code wasn't using memstore, i.e. hitting disk, hence the slowdown

satazor commented:
I've made the changes. On my end, this is good to go but please test it and inspect both a log and entry hashes in https://explore.ipld.io/ to see how it looks like!

aphelionz commented on src/entry.js:
```diff
@@ -90,22 +90,21 @@ class Entry {
       id: entry.id,
       payload: entry.payload,
-      next: entry.next,
+      next: entry.next.map(cid => ({ '/': cid })),
```
@satazor @haadcode @shamb0t  Might we be able to approach backwards compatibility here by first checking for the cid and then checking for `entry.next` as a fallback?

satazor replied:
Yes we can indeed do that. Will change!

satazor replied:
I think it’s worth creating a helper in utils/ just for that, what do you think? The helper would read and write dag nodes, with an argument indicating which properties need to be linked.

aphelionz replied:
Love it. Let's see how it looks!

satazor replied:
@aphelionz I've created `toIpldNode` and `fromIpldNode` in the utils. Can you check how this looks?

aphelionz commented on src/entry.js:
```diff
@@ -122,21 +121,21 @@ class Entry {
-      clock: new Clock(data.clock.id, data.clock.time)
+      id: dagNode.id,
+      payload: dagNode.payload,
+      next: dagNode.next.map((cid) => cid.toBaseEncodedString()),
```
@satazor @haadcode @shamb0t Likewise here.

aphelionz commented on test/utils/mem-store.js:
```diff
@@ -17,44 +19,56 @@ const createMultihash = (data, hashAlg) => {
-// const LRU = require('lru')
-// const ImmutableDB = require('./immutabledb-interface')
-// const createMultihash = require('./create-multihash')
+const transformLinksIntoCids = (data) => {
```
We probably want to move this to utils.js

satazor replied:
Hmm, this is not needed anywhere else for now. Perhaps we shall move it if it’s necessary in more than one place?

aphelionz replied:
Works for me.

aphelionz commented on API.md:
```diff
@@ -78,28 +78,28 @@ log.append({ some: 'data' })
 //     id: 'A',
 //     payload: 'text',
 //     next: [ 'QmV1KFxZnaguPFp57PMXz5Dd1tf6z9U8csJQoy4xWDzzts' ],
 //     v: 0,
```
Additionally, we should probably increment this from 0 to 1 as well. See the thread and comment here: https://github.com/orbitdb/ipfs-log/issues/211#issuecomment-450730115

cc @haadcode and @shamb0t

satazor replied:
Yep, will change.

shamb0t replied:
Thank you!

aphelionz commented:
Finally, run `make rebuild` and push all built files and es5 files when you make the last commit :)

RichardLitt commented:
> Finally, run `make rebuild` and push all built files and es5 files when you make the last commit :)

Should we add this to the Contributing file?

aphelionz commented:
@RichardLitt Yes, definitely. Thanks for catching that.

shamb0t commented on src/utils/from-ipld-node.js:
```diff
@@ -0,0 +1,28 @@
+}
+
+const fromIpldNode = (dagNode, props) => {
+  const obj = Object.assign({}, dagNode)
```
@satazor version 0 of ipfs-log appears to create 'dag-pb' objects so this still fails backwards-compatibility when I try to open an old orbit-db address. I'm not sure its the best way but it works if we change this to sth like the following: `const obj = dagNode.data ? Object.assign({}, JSON.parse(dagNode.data)) : Object.assign({}, dagNode)`. What do you think?

That along with the version increment @aphelionz mentioned should do the trick

satazor replied:
You are right absolutely right. I will take care of that and add tests as well.

satazor replied:
One question: If I call `fromMultihash` with a hash of a `dag-pb` and then call `toMultihash`, it should give the same hash right? If so, I think it's worth doing "proper" migration code that handles reading/writing nodes with version support.

aphelionz replied:
How would that work in a "mixed" log - i.e. one with some `v: 0` entries and some `v: 1` entries? Another way of asking is - how would the log know which multihash type it was created from?

satazor replied:
@aphelionz The way I see it:

- Have a `read` step given a multihash. The read function would essentially return a JSON object directly from `dag.get`. This is where we can test if a hash is a `dag-pb` and use JSON.parse or not. We may use [cid](https://github.com/multiformats/js-cid) for this or just a `typeof` check.
- Have a migration pipeline. We setup a set of functions to migrate from previous versions, e.g.: we would have functions to migrate from v0 to v1, v1 to v2 and so on. I suspect we may find a few libraries on npm that may help us with this. This is based on the `v` field that is tagged on the JSON object.
- Finally, the `fromIpldNode` and `toIpldNode` would still be called by `entry-io.js` and `log.js` but is out of scope of the "migration" pipeline.

shamb0t replied:
Hmm good point, I agree fromMultihash and toMultihash should return the same. Perhaps we should change `Entry.toMutihash` and use `DAGNode.create` and return the cid if `entry.v === 0`. What do you think?

satazor replied:
@shamb0t Yes I can do that but it feels ad-hoc. I would suggestion embrancing/crafting a migration pipeline that would take care of that and future changes on both reads and writes. What do you think?

shamb0t replied:
Woops missed your last comment. Migration pipeline would be great but is it not an issue that each entry needs to be signed? That means logs with many users wouldnt be able to be migrated bc incrementing the version of an entry would require new signature from each user. So I think backwards compatibility will just be for reading old entries but not updating them. Does this make sense? I may be missing something

shamb0t replied:
@satazor yea it is quite hacky but not sure we can update existing entries as they are signed

satazor replied:
@shamb0t Correct, new entries would be always written using the latest version they are running. Note that I may still a node running an old version and I will be writing those nodes in that version.

Nevertheless, calling `Entry.fromMultihash` followed by `toMultihash` would respect the `version` the entry is written, thus giving the same hash. But calling, `Entry.create` would be creating an entry with the most recent version always.

satazor replied:
> @satazor yea it is quite hacky but not sure we can update existing entries as they are signed

Yep, we shouldn't update the existing entries. The migration I'm describing is on reads, thus is like a `runtime` migration without affecting the storage. The migration should also read writes only with the goal of giving the same multihash when calling `Entry.fromMultihash` followed by toMultihash`

satazor replied:
(updated my last comment)

shamb0t replied:
Ah maybe I am misunderstanding what you mean by migrating here? Not updating existing entries?

satazor replied:
One important thing: if a log instance is trying to read/write a node in which the version is not supported, it should fail with an error. This is the case where an outdated instance is being synched with a newer one.

satazor replied:
Perhaps "migration" is a bad name because it somewhat implies changing the storage. Do you have a better name?

shamb0t replied:
> Perhaps "migration" is a bad name because it somewhat implies changing the storage. Do you have a better name?

I guess not :p I think we're in agreement about desired behaviour but not sure Im grasping where this step should be and how its different to fromMultihash. What do you mean by migrate from v0 to v1 here?

> * Have a migration pipeline where we setup a set of functions to migrate from previous versions, e.g.: we would have functions to migrate from v0 to v1, v1 to v2 and so on.

shamb0t replied:
Do you mean fromMultihash should be called within this pipeline and not do the version check itself?

satazor replied:
> I guess not :p I think we're in agreement about desired behaviour but not sure Im grasping where this step should be and how its different to fromMultihash. What do you mean by migrate from v0 to v1 here?

I went to the whiteboard and wrote something to help you (and me):

![image](https://user-images.githubusercontent.com/1017236/50649185-1df9a100-0f75-11e9-8815-6e3ee6f62f81.png)

The "writing" would be similar but backwards.

This procedure will allows us to change fields or the data-structure of the entries easily. It would also allow us to change the codec in the future, in case we are not happy with `dag-cbbor`. While this is not complex to code, perhaps I suggest to keep this PR simple and and address this on #211.

I've changed the code to handle `dag-pb` nodes, will add a few tests.

satazor replied:
Expanding on this, the `migration` pipeline would be composed by something like this:

```js
{
    '2': {
        up(data) { return transformedData },
        down(data) { return transformeData },
        normalizeIpldLinks(data) { return dataWithLinksNormalizedToMultihashes },
     },
     '3': ...
}
```

Where:

- `up()` would transform the JSON `data` to their next version
- `down` would transform the JSON `data` to their previous version
- `normalizeIpldLinks` would be called by the step `1` (just once) so that IPLD links are transformed to regular multihashes (strings)


}

satazor replied:
> While this is not complex to code, perhaps I suggest to keep this PR simple and and address this on #211.

While I said this, the `fromMultihash` and `toMultihash` is still a problem in this PR. The migration scheme I detailed above would solve this or we may do a check like @shamb0t suggested in the `toMultihash` functions for now.

aphelionz replied:
We are very close, @satazor! I would like @hardcore to weigh in but I would personally implore you to finish that last piece and achieve full backwards compatability.

Alternatively, it might be ok if fromMultihash _accepts_ either type but toMultihash only returns cids.

satazor replied:
@aphelionz fromMultihash isn't the problem, toMultihash is. We could add a optional `version` argument to it, but it would feel strange, specially since the entries do not need it. I will look into what needs to be changed to add a `v` field to the `log.js` and `log-io.js` but I'm afraid it would require a few changes, that's why I suggested postponing adding support for IPLD in them later on.

Lets see what @haadcode has to say.

haadcode replied:
Thanks for the ping 👍 Reading through the conversation, couple of thoughts come to mind:
- We should try to keep the solution as simple (=lightweight) as possible for now
- I like the idea of a "migration pipeline" as this is prolly not the last time we'll need to provide migrations
- That ^ said, I think it would be helpful to spend more time to figure out the full solution to make sure it's the right solution. it sounds to me this can be done at a later point and doesn't have to be part of this PR. It also reminds me that we'll probably need something similar, sooner or later, for the stores in OrbitDB and ideally a solution would facilitate both use cases (log and stores).
- We should (try to) keep all "log information" in the entries: this is something we looked into very early on and decided to not add "meta info" for the log itself and rather consider the structure of the log to be defined by the entries + public API of the log (ie. the functions you can call). This keeps the abstractions and data structures simple imo.
- We should, if possible, make it so that a log can contain mixed (Entry) versions. Ie. read a v0 log, but when appending to it, it'll be in v1 format, eg. `(e1, v0)->(e2, v0)->(e3, v1)`.
- Re. the `to/FromMultihash`, agreed that the semantics should be consistent and it is indeed a bit tricky: on one hand it would make sense that `read` (=fromMultihash) is backwards compatible in that it can take either a multihash or CID, and that `write` (=toMultihash) would return the same hash. On the other hand, it would make sense that `write` returns the _same kind of hash for every log_ and given we're going from multihashes to CIDs in the entries, it would make sense to me that the `write` returns a CID.

It sounds to me that the big question is whether, when persisting a log to IPFS, should return a multihash or a CID. The options are to 1) write a full-fledged migration pipeline or 2) figure out the right semantics for `to/fromMultihash` between the different versions. I think there's a third option, which would be to add `toCID` and `fromCID` functions that semantically do the same as `to/fromMultihash` but work on CIDs. Further, we could add a `console.warn("This function will be deprecated in the next release")` to `to/fromMultihash` to encourage users to switch. This would probably still require us to provide backwards compat in the `fromCID` part (maybe?), but from semantics perspective it would be consistent. 

What do you think, would that be a good approach and would it cover all the cases discussed above?

satazor replied:
> We should try to keep the solution as simple (=lightweight) as possible for now

Agree.

> I like the idea of a "migration pipeline" as this is prolly not the last time we'll need to provide migrations
That ^ said, I think it would be helpful to spend more time to figure out the full solution to make sure it's the right solution. it sounds to me this can be done at a later point and doesn't have to be part of this PR. It also reminds me that we'll probably need something similar, sooner or later, for the stores in OrbitDB and ideally a solution would facilitate both use cases (log and stores).

Agree, I can later on formalize my proposal in #211 

> We should (try to) keep all "log information" in the entries: this is something we looked into very early on and decided to not add "meta info" for the log itself and rather consider the structure of the log to be defined by the entries + public API of the log (ie. the functions you can call). This keeps the abstractions and data structures simple imo.

> Re. the to/FromMultihash, agreed that the semantics should be consistent and it is indeed a bit tricky: on one hand it would make sense that read (=fromMultihash) is backwards compatible in that it can take either a multihash or CID, and that write (=toMultihash) would return the same hash. On the other hand, it would make sense that write returns the same kind of hash for every log and given we're going from multihashes to CIDs in the entries, it would make sense to me that the write returns a CID.

Making `toMultihash` always return a CID would indeed simplify. Is there any reason, at this point, to have them consistent? Will it somehow make the log behave badly? Looking at the code, it seems it doesn't. As I said. making `fromMultihash` consistent for entries is easy, but not for logs because we would have to tag them with the version as well. We got 3 options:

1. Drop the consistency of `fromMultihash` and `toMultihash`; the `toMultihash` would always use the latest codec (dag-cbor) and return a cid (is still a string but more complete)
2. Add a `version` argument to the Log constructor that defaults to the latest version. The `fromMultihash` function would create a Log instance, passing `version` = `data.version || 0`. Then, the `toMultihash` would have the version to decide to write a `dag-pb` or a `dag-cbor`.
3. When calling `toMultihash`, analyse the head entries in order to calculate the higher one. This would be used to decide which codec to use. But I think this would require further IO?

What do you suggest going for?

>  I think there's a third option, which would be to add toCID and fromCID functions that semantically do the same as to/fromMultihash but work on CIDs. Further, we could add a console.warn("This function will be deprecated in the next release") to to/fromMultihash to encourage users to switch.

I like this suggestion, perhaps we could do that in another PR after we figure this out.

shamb0t replied:
Thanks for the whiteboarding/explanation @satazor, very helpful and I see this is more future-proof. With that, I agree we should try to keep this PR as simple as possible. I've successfully opened an old orbit-db address using ipfs-log from this PR :tada: if we agree to address the migration pipeline separately, is this ready for final review/merge?

@haadcode re: "mixed-logs" it makes sense that appending would use the latest version, but we should discuss how a peer knows it needs to update....it will only know once it receives an entry for a version it doesnt understand, and as @satazor said it should error out in such case, before receiving such an entry it would still be able to append with a previous version

satazor replied:
@shamb0t 

> I've successfully opened an old orbit-db address using ipfs-log from this PR 🎉 if we agree to address the migration pipeline separately, is this ready for final review/merge?

Not yet, we are still deciding if calling `fromMultihash` followed by a `toMultihash` should be consistent or not. Based on the decision, I might need to do some changes and add a few tests.

satazor replied:
@shamb0t I forgot to mention that solution as well, I updated my comment above and changed option 3 to be to/fromCID.

aphelionz replied:
I like introducing to/fromCID. It's not _so_ much work to change or remove it in the future, and then we can keep to/fromMultihash around. All four functions will do what they say - unless there's a more specific term we can name to/fromMultihash to that I'm not thinking of right now.

satazor replied:
If we go for the `to/fromCID` route, should we change every occurrence of hash, multihash to cid in all the source code and tests? If so, we can't change the `hash` property to `cid` in entries, otherwise the signatures would be invalid. There's also `Log.findTailHashes` but I think it's not part of the public API, so we can rename it.

satazor replied:
@haadcode could you clarify if `log.fromEntry` and `findTailHashes` are part of the public API?

haadcode replied:
I believe `fromEntry` is but `findTailHashes` is not.

satazor replied:
Ooops, I mean `log.fromEntryHash`.

haadcode replied:
`log.fromEntryHash` I don't think is part of the public API

shamb0t replied:
@satazor I feel changing all occurrences is unnecessary unless we do so in the whole codebase, entry included or it would be confusing. This would be quite a lot of work imo. Do you think its confusing to return a multihash from the `hash` property? The multihash from the cid is still what we're storing right? Or do I misunderstand?

satazor replied:
Yes the `hash` will have the cid from now on while old entries still have the multihash. Both are strings and reading will be backwards compatible.

satazor replied:
@shamb0t I agree with you. We can change `hash` to `cid` and maintain compatibility, including signatures verification. Moreover, we can keep the `hash` property around, but hidden, by making it non-enumerable.

satazor replied:
I'm give `toCID` and `fromCID` a shot, see my comment below.

satazor commented:
So I took the journey of adding `toCID` and `fromCID` as well as replace all `hash` occurrences with `cid`. Everything is working as expected and added a few tests to cover the v0 compatibility.

In order to do this properly I had to dig into the code and I detected some inconsistencies:

- A lot of JSDoc were wrong or outdated; I fixed the wrong ones and updated the outdated ones.
- `log.heads` return cids, `log.tails` return entries and `log.tailCids` (old `log.tailHashes`) return cids. The most correct would be `log.heads` and `log.tails` return cids and `log.headEntries` and `log.tailEntries` to return entries.
- I noticed the the `exclude` argument used in several places were mean to be an `Array<string,Entry>` but in fact the code only handles `Array<string>`. Moreover, there's either a bug or a wrong test regarding this: `Log.fromEntry` passes the `exclude` as cids to `LogIO.fetchAll` which in turn calls `LogIO.fetchAll`. If you console.log the `cache` after parsing the `excluded` argument, we get `{ undefined: <cid> }`. I think the code is wrong.. I tried to fix it by just passing the entries but some related tests started to failing.
- API.md and README.md are outdated. I haven't updated those because I think we should merge #177 first.
- Examples are not working and seem outdated. Again,  I haven't updated those because I think we should merge #177 first.

Please advise regarding the points I mentioned before.

-------------------------

Please review the PR to see how it looks like. @shamb0t it would be great if you could test it locally, please :D.

satazor commented on src/entry.js:
```diff
@@ -23,172 +29,226 @@ class Entry {
     if (!entry.key) throw new Error("Entry doesn't have a key")
     if (!entry.sig) throw new Error("Entry doesn't have a signature")
 
-    const e = Object.assign({}, {
```
Creating a new object was unnecessary, therefore I've removed the shallow clone.

satazor commented on src/entry.js:
```diff
@@ -23,172 +29,226 @@ class Entry {
       payload: entry.payload,
       next: entry.next,
       v: entry.v,
-      clock: new Clock(entry.clock.id, entry.clock.time)
```
Correct if I'm wrong, but this was unnecessary.

satazor commented on src/log-io.js:
```diff
@@ -53,19 +75,34 @@ class LogIO {
     }
   }
 
-  static async fromEntryHash (ipfs, entryHash, id, length = -1, exclude, onProgressCallback) {
```
Should we keep this but deprecate it? I'm not sure if it's part of the public API.

haadcode replied:
I believe we can deprecate this. @shamb0t can you confirm?

satazor replied:
@haadcode yes I must restore the function and declare it as deprecated.

satazor commented on src/log-io.js:
```diff
@@ -53,19 +75,34 @@ class LogIO {
+    if (!isDefined(entryCid)) throw new Error("'entryCid' must be defined")
 
-    // Convert input hash(es) to an array
-    const entryHashes = Array.isArray(entryHash) ? entryHash : [entryHash]
```
This is the code I think it's bogus. This will cause the EntryIO `cache` object to be  `{ undefined: 'zd..' }`

satazor replied:
Solved by @shamb0t

shamb0t commented:
Thanks @satazor will take a look at this today!

shamb0t commented:
Thanks for this @satazor! I honestly feel while this is the right direction, further changes are required for the hash/cid replacements and is perhaps out of scope for this PR. 

> * `log.heads` return cids, `log.tails` return entries and `log.tailCids` (old `log.tailHashes`) return cids. The most correct would be `log.heads` and `log.tails` return cids and `log.headEntries` and `log.tailEntries` to return entries.

This sounds reasonable to me.

> * I noticed that the `exclude` argument used in several places were meant to be an `Array<string,Entry>` but in fact the code only handles `Array<Entry>`. Moreover, there's either a bug or a wrong test regarding this: `Log.fromEntry` passes the `exclude` as cids to `LogIO.fetchAll` which in turn calls `LogIO.fetchAll`. If you console.log the `cache` after parsing the `excluded` argument, we get `{ undefined: <cid> }`. I think the code is wrong.. I tried to fix it by just passing the entries but some related tests started to failing.

This indeed looks like an issue with transforming the exclude<entries> to exclude<entry.hashes>, cache[e.cid] = e gives { 'undefined': somehash} bc e is a hash. Looks to me like the test is wrong but the test description is strange ("included excluded entries?") I suspect the test should expect to not include that entry, @haadcode can you confirm?

> * The API.md and README.md are outdated. I haven't updated those because I think we should merge #177 first.
> * The examples are not working and seem outdated. Again, I haven't updated those because I think we should merge #177 first.

Yes agreed we can leave the documentation until merging that PR :+1: 

> * I've renamed `Log.fromEntryHash` to `Log.fromEntryCid`. Should I restore `Log.fromEntryHash` and deprecate it? I've removed it but I'm not sure if it's part of the public API or not. It's mentioned in the `README.md` but not in the `API.md`. My opinion is that `Log.fromEntryCid` makes more sense than `Log.fromEntry` in terms of public API, which leads to the next point.

> * Currently, the Log constructor and the `Log.fromEntry` functions are not backwards compatible in case you pass a v0 entry object. We need to transform the entries to their latest version in order to make it so (hence the migration pipeline, but we may hack it for now by just changing `hash` to `cid`). Though, _I think that the entry objects are an implementation detail that should be hidden_ from consumers of the Log instance. Therefore, I strongly believe that we should not return entries from within most functions of the log but instead rely on their cids. Anyway, this would require a lot of changes so it's out of question, but let me know what you think about the backwards compatibility here.

This does indeed break backwards compatibility and also requires changing occurrences of `hash` in the rest of the orbit-db codebase. Also `Log.fromEntryHash/Cid` is used in `orbit-db-store` and should be returned.  I agree that the implementation details should be hidden. I think this should be out of the scope of this PR and imo would be simpler and preferable to either forego consistency bw to/fromMultihash or hack toMultihash to give the correct cid by checking the entry version and using dag-pb or dag-cbor according to that. I do think we should move in that direction but will require more thought and should be a separate PR imo. What do you think?

haadcode commented on src/entry.js:
```diff
@@ -23,172 +29,226 @@ class Entry {
-      next: data.next,
-      v: data.v,
-      clock: new Clock(data.clock.id, data.clock.time)
+      cid,
```
What's happening here? What's the key, what's the value?

shamb0t replied:
Both key and value are cid (this breaks backwards compatibility and should be `hash` if dagNode.read returns an entry with e.v === 0

satazor replied:
`cid` are both the key and the value, this is a feature of ES6 I think which is supported since node 4, see: https://node.green/#ES2015-syntax-object-literal-extensions-shorthand-properties

satazor replied:
Also, it replaces the old `hash` for the new fromCID/toCID.

satazor commented on src/entry.js:
```diff
@@ -23,172 +29,226 @@ class Entry {
 
     // Ensure `entry` follows the correct format
     const e = {
       hash: null,
```
@haadcode toMultihash should be backwards compatible, therefore it must create the same mulithash than before. On the contrary, toCID returns a new cid.

satazor commented:
@shamb0t thanks for the response

Regarding:

> > log.heads return cids, log.tails return entries and log.tailCids (old log.tailHashes) return cids. The most correct would be log.heads and log.tails return cids and log.headEntries and log.tailEntries to return entries.

> This sounds reasonable to me.

Should I create an issue for this? Since this is a breaking change I guess we should take care of this later.

> This indeed looks like an issue with transforming the exclude to exclude<entry.hashes>, cache[e.cid] = e gives { 'undefined': somehash} bc e is a hash. Looks to me like the test is wrong but the test description is strange ("included excluded entries?") I suspect the test should expect to not include that entry, @haadcode can you confirm? Okay the code is indeed wrong, will make a PR to fix

I've rebased against master, thanks for looking and fixing this.


> This does indeed break backwards compatibility and also requires changing occurrences of hash in the rest of the orbit-db codebase. Also Log.fromEntryHash/Cid is used in orbit-db-store and should be returned. I agree that the implementation details should be hidden. I think this should be out of the scope of this PR and imo would be simpler and preferable to either forego consistency bw to/fromMultihash or hack toMultihash to give the correct cid by checking the entry version and using dag-pb or dag-cbor according to that. I do think we should move in that direction but will require more thought and should be a separate PR imo. What do you think?

To have 100% backwards compatible we can:

- Add the `hash` property back but make it non-enumerable by using `Object.defineProperty` when creating entries. This way, `orbit-db` and other consumers of this module would still be working.
- Restore `fromEntryHash` because `orbit-db` is using it
- Ensure that the "imported" entries get migrated to the latest version, e.g.: in `Log.fromEntry`

All these 3 points are very easy to do. Let me know if I'm missing something and if you still prefer to rollback all the `hash` to `cid` changes in this PR. @haadcode could you also provide your feedback?

shamb0t commented:
> Should I create an issue for this? Since this is a breaking change I guess we should take care of this later.

Agreed we should take care of this later, opened an issue [here](https://github.com/orbitdb/ipfs-log/issues/218) :+1: 

> I've rebased against master, thanks for looking and fixing this.

Thank you! :+1: 

> 
> To have 100% backwards compatible we can:
> 
> * Use `cid` in entries and add `hash` property back but make it non-enumerable by using `Object.defineProperty`. This way, `orbit-db` and other consumers of this module would still be working if they access `entry.hash`.
> * Restore `fromEntryHash` (but deprecated) because `orbit-db` is using it
> * Ensure that the "imported" entries get migrated to the latest version, e.g.: when using `Log.fromEntry`
> 
> All these 3 points are very easy to do. Let me know if I'm missing something and if you still prefer to rollback all the `hash` to `cid` changes in this PR. @haadcode could you also provide your feedback on this?

These sound good, I would prefer moving forward than rolling back the changes as I think this is the correct direction and is useful. Am I correct in understanding the above would keep consumers of this module from breaking but will not read newly created entries until all `hash` occurrences are changed to `cid`. (assuming we ensure that 'imported' entries get migrated, then calling `entry.cid` in consumers will have the same affect for old entries and support new entries also) :+1: 

We can probably keep fromEntryCid as long as we still have the `hash` property in entries, this is also a simple change in orbit-db-store

haadcode commented on test/entry.spec.js:
```diff
@@ -214,14 +350,14 @@ Object.keys(testAPIs).forEach((IPFS) => {
           err = e
         }
-        assert.strictEqual(err.message, 'Invalid hash: undefined')
+        assert.strictEqual(err.message, 'Invalid multihash: undefined')
```
This prolly should be "Invalid CID: undefined"?

satazor replied:
Since the function is `fromMultihash`, I decided to keep it like so. Let me know if you want this changed.

satazor replied:
Since `fromMultihash` is a proxy to `fromCID` now, it's now `Invalid CID` like you said.

haadcode commented on test/entry.spec.js:
```diff
@@ -187,21 +235,109 @@ Object.keys(testAPIs).forEach((IPFS) => {
+        assert.strictEqual(final.cid, expectedMultihash)
+      })
+
+      it('creates a entry from ipfs multihash of v0 entries', async () => {
```
Excellent, great to have a test for this! 👍

satazor replied:
Thanks!

haadcode commented on test/entry.spec.js:
```diff
@@ -187,21 +235,109 @@ Object.keys(testAPIs).forEach((IPFS) => {
+
+    describe('fromMultihash', () => {
+      it('creates a entry from ipfs multihash', async () => {
+        const expectedMultihash = 'QmRyrbupohhmSoXxQ56XthEiK6YEqK5ZN8SysJdo8GGdHq'
```
Great to have a test for this! 👍

satazor replied:
Awesome1

haadcode commented:
It all looks good to me, but having said that, I haven't worked on this at all and it's hard to confirm for sure on individual implementation details. From what I can tell, it looks correct.

I would want to request one final test though: we should have an old log (as fixtures prolly) and make sure it can be opened and operated on with the new version (this PR) in all possible ways: fromCID/toCID/fromMultihash/toMultihash/etc. We should make sure that the entries are indeed old version to make sure it can handle them correctly, ie. instead of `identity` the entries have `key` and instead of `cid` they have `hash` etc. I believe having such test would make everyone more comfortable to land this knowing that we don't break everything for users with old log. And we should have the same test in OrbitDB to test the compatibility (out of scope for this PR, but still). How does that sound @shamb0t and @satazor?

You've done great job @satazor and @shamb0t, thank you for doing all this! ❤️

satazor commented:
I've made the changes mentioned above in this last commit. Also I changed `toMultihash/fromMultihash` to simply call `toCID/fromCID` so that it's simpler to reason about the code and tests.

I haven't added interop tests for `Log.fromEntry` and `Log.fromJSON` because I couldn't find any tests for them. Should I leave the PR like this or add a few? If so, where do I add them?

satazor commented:
@shamb0t @aphelionz @haadcode this is ready to be reviewed. Some things to point out:

- `Entry.toMultihash` returns the same result than the current code on master, while the new `Entry.toCID` returns a CID based based on `cbor` and with IPLD links; the same applies to `Log.toMultihash` and `Log.toCID`
- `Entry.fromMultihash` and the new  `Entry.fromCID`  both able to read old and new entries; the same applies to `Log.fromMultihash` and `Log.fromCID`
- All the entry objects passed in as arguments to functions in the public API are interoperable. The same applies to entries returned from the public API. This is thanks to the `Entry.ensureInterop` function.
- I haven't added interop tests for `Log.fromEntry` and `Log.fromJSON` because I couldn't find any tests for them. Should I leave the PR like this or add a few? If so, where do I add them?

shamb0t commented:
Thank you @satazor! Everything lgtm :+1: I've tested and it works with old orbit-db addresses. I'm happy to merge this without additional interop tests, I'll add some to the orbit-db tests too.

We can discuss steps re: migration pipeline in #211 looks to me that a lot of the preliminary work is done here which is great!

satazor commented:
@shamb0t thanks for looking into this! Regarding the documentation, you will be updating it as part of #177?

shamb0t commented:
@satazor yes I've added those to the API documentation in #177
